### ISSUE TITLE
Discussion: distinguish KeyIsLocked error and other errors

### ISSUE THREAD

**Issue by breeswish** [tikv/tikv#3481]
# Discussion

Currently there is something unsound in mvcc readers, caused by `KeyIsLocked` error. `KeyIsLocked` is a kind of recoverable mvcc errors that the scanner should be able to proceed when facing it. Meanwhile, other kind of mvcc errors can be considered to be not recoverable, for example, the error produced by failing to decode ts (usually means a data corruption) or the error that given key is not in current region's range (RegionError). For these non-recoverable errors, we should either panic, or throw errors to the user and abort the operation.

Currently we mixed these kind of errors together in bottom layers:

- In the bottom most, `check_lock` returns a `Result<>`, which mixed KeyIsLocked error and other errors.
- In `MvccReader::get`, we don't distinguish them, also returns a union `Result<>`.
- In `StoreScanner::scan`, we start to distinguish them with the type `Result<Vec<Result<KvPair>>>`, where the outer `Result` is non-recoverable errors and the inner `Result` is recoverable errors.

This works mostly, until now we refined mvcc reader scanners: In new scanners, when we meet recoverable errors, we have some remaining steps to do to prepare for next access. However for non-recoverable errors this is unnecessary: currently we treat them both as recoverable errors when calling `check_lock` and prepare for next access even for non-recoverable errors.

Some ideas to solve this edge case.

### Solution 1

Extract KeyIsLocked from errors, so that we are forced to treat them differently.

1. `check_lock` returns `Result<LockInfo<..>>` instead of `Result<..>`, so that KeyIsLocked are extracted.
2. MvccReader scanners return `Result<Either<LockInfo, (Key, Value)>>`.
3. Naturally, storeScanner returns `Result<Vec<Either<LockInfo, KvPair>>>`.

### Solution 2

Make a new error kind, for example, `RecoverableError`, in which currently there is only 1 error, which is, KeyIsLocked.

Compare to solution 1, this solution reserves space so that in future there might be more kind of recoverable errors.

### Solution 3

Distinguish KeyIsLocked and other error in the new MvccReader scanners. However the code might become a little verbose.

Thoughts are welcome! @BusyJay @zhangjinpeng1987 @MyonKeminta

**Comment by BusyJay** [comment]
Is `KeyIsLock` always recoverable? For example, in coprocessor scan, when a key is locked, I think it should be aborted immediately.

**Comment by breeswish** [comment]
@BusyJay Right. In coprocessor (the caller) we don't need to recover it however for the underlying interface it still needs to be recoverable.

**Comment by breeswish** [comment]
@BusyJay Our KV interface requires it to be recoverable :( It should continues scanning even if there are locks.
